compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2552,6 +2554,50 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
             _ => {}
         }
     }
+
+    fn suggest_derive(&self, err: &mut Diagnostic, trait_pred: ty::PolyTraitPredicate<'tcx>) {
+        let Some(diagnostic_name) = self.tcx.get_diagnostic_name(trait_pred.def_id()) else {
+            return;
+        };
+        let (adt, substs) = match trait_pred.skip_binder().self_ty().kind() {
+            ty::Adt(adt, substs) if adt.did().is_local() => (adt, substs),
+            _ => return,
+        };
+        // Ensure all fields impl the trait.
+        for variant in adt.variants() {
+            for field in &variant.fields {
+                let obligation = self.mk_trait_obligation_with_new_self_ty(
+                    self.tcx.param_env(adt.did()),
+                    trait_pred,
+                    field.ty(self.tcx, substs),
+                );
+                if !self.predicate_must_hold_modulo_regions(&obligation) {
+                    return;
+                }
+            }
+        }
+        let can_derive = match diagnostic_name {
+            sym::Default => !adt.is_enum(),
+            sym::PartialEq | sym::PartialOrd => {
+                let rhs_ty = trait_pred.skip_binder().trait_ref.substs.type_at(1);
+                trait_pred.skip_binder().self_ty() == rhs_ty
+            }
+            sym::Eq | sym::Ord | sym::Clone | sym::Copy | sym::Hash | sym::Debug => true,
+            _ => false,
+        };
+        if can_derive {
+            err.span_suggestion_verbose(
+                self.tcx.def_span(adt.did()).shrink_to_lo(),
+                &format!(
+                    "consider annotating `{}` with `#[derive({})]`",
+                    trait_pred.skip_binder().self_ty().to_string(),
```

Comment:
nit: `ToString::to_string` is not needed here I think, or below.
```suggestion
                    trait_pred.skip_binder().self_ty(),
```